File: push/src/main/java/com/xiaomi/push/service/PushClientsCollection.java
package com.xiaomi.push.service;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

final class PushClientsCollection {
    private final ConcurrentHashMap<String, HashMap<String, PushClientsManager.ClientLoginInfo>> activeClients = new ConcurrentHashMap<>();
    private final List<PushClientsManager.ClientChangeListener> clientChangeListeners = new ArrayList();

    void addActiveClient(PushClientsManager.ClientLoginInfo clientLoginInfo) {
        HashMap<String, PushClientsManager.ClientLoginInfo> map = this.activeClients.get(clientLoginInfo.chid);
        if (map == null) {
            map = new HashMap<>();
            this.activeClients.put(clientLoginInfo.chid, map);
        }
        map.put(PushClientsManager.getSmtpLocalPart(clientLoginInfo.userId), clientLoginInfo);
        notifyListeners();
    }

    void addClientChangeListener(PushClientsManager.ClientChangeListener clientChangeListener) {
        this.clientChangeListeners.add(clientChangeListener);
    }

    void deactivateAllClientByChid(String str) {
        HashMap<String, PushClientsManager.ClientLoginInfo> map = this.activeClients.get(str);
        if (map != null) {
            Iterator<PushClientsManager.ClientLoginInfo> it = map.values().iterator();
            while (it.hasNext()) {
                it.next().unwatch();
            }
            map.clear();
            this.activeClients.remove(str);
        }
        notifyListeners();
    }

    void deactivateClient(String str, String str2) {
        HashMap<String, PushClientsManager.ClientLoginInfo> map = this.activeClients.get(str);
        if (map != null) {
            PushClientsManager.ClientLoginInfo clientLoginInfo = map.get(PushClientsManager.getSmtpLocalPart(str2));
            if (clientLoginInfo != null) {
                clientLoginInfo.unwatch();
            }
            map.remove(PushClientsManager.getSmtpLocalPart(str2));
            if (map.isEmpty()) {
                this.activeClients.remove(str);
            }
        }
        notifyListeners();
    }

    int getActiveClientCount() {
        return this.activeClients.size();
    }

    Collection<PushClientsManager.ClientLoginInfo> getAllClientLoginInfoByChid(String str) {
        if (this.activeClients.containsKey(str)) {
            return ((HashMap) this.activeClients.get(str).clone()).values();
        }
        return new ArrayList();
    }

    ArrayList<PushClientsManager.ClientLoginInfo> getAllClients() {
        ArrayList<PushClientsManager.ClientLoginInfo> arrayList = new ArrayList<>();
        Iterator<HashMap<String, PushClientsManager.ClientLoginInfo>> it = this.activeClients.values().iterator();
        while (it.hasNext()) {
            arrayList.addAll(it.next().values());
        }
        return arrayList;
    }

    PushClientsManager.ClientLoginInfo getClientLoginInfoByChidAndUserId(String str, String str2) {
        HashMap<String, PushClientsManager.ClientLoginInfo> map = this.activeClients.get(str);
        if (map == null) {
            return null;
        }
        return map.get(PushClientsManager.getSmtpLocalPart(str2));
    }

    Iterable<HashMap<String, PushClientsManager.ClientLoginInfo>> getActiveClientMaps() {
        return this.activeClients.values();
    }

    void removeActiveClients() {
        Iterator<PushClientsManager.ClientLoginInfo> it = getAllClients().iterator();
        while (it.hasNext()) {
            it.next().unwatch();
        }
        this.activeClients.clear();
    }

    void removeAllClientChangeListeners() {
        this.clientChangeListeners.clear();
    }

    void removeClientChangeListener(PushClientsManager.ClientChangeListener clientChangeListener) {
        this.clientChangeListeners.remove(clientChangeListener);
    }

    private void notifyListeners() {
        Iterator<PushClientsManager.ClientChangeListener> it = this.clientChangeListeners.iterator();
        while (it.hasNext()) {
            it.next().onChange();
        }
    }
}

File: legacy/src/main/java/com/xiaomi/push/service/PushClientsCollection.kt
package com.xiaomi.push.service
import io.github.magisk317.mipush.protocol.model.*

import java.util.concurrent.ConcurrentHashMap

class PushClientsCollection {
    private val activeClients =
        ConcurrentHashMap<String, HashMap<String?, PushClientsManager.ClientLoginInfo>>()
    private val clientChangeListeners = ArrayList<PushClientsManager.ClientChangeListener>()

    fun addActiveClient(clientLoginInfo: PushClientsManager.ClientLoginInfo) {
        val clientsByChannel = activeClients.getOrPut(clientLoginInfo.chid) { HashMap() }
        clientsByChannel[smtpLocalPart(clientLoginInfo.userId)] = clientLoginInfo
        notifyListeners()
    }

    fun addClientChangeListener(clientChangeListener: PushClientsManager.ClientChangeListener) {
        clientChangeListeners += clientChangeListener
    }

    fun deactivateAllClientByChid(chid: String) {
        activeClients[chid]?.let { clientsByChannel ->
            clientsByChannel.values.forEach { it.unwatch() }
            clientsByChannel.clear()
            activeClients.remove(chid)
        }
        notifyListeners()
    }

    fun deactivateClient(chid: String, userId: String) {
        activeClients[chid]?.let { clientsByChannel ->
            clientsByChannel[smtpLocalPart(userId)]?.unwatch()
            clientsByChannel.remove(smtpLocalPart(userId))
            if (clientsByChannel.isEmpty()) {
                activeClients.remove(chid)
            }
        }
        notifyListeners()
    }

    fun getActiveClientCount(): Int = activeClients.size

    fun getAllClientLoginInfoByChid(chid: String): Collection<PushClientsManager.ClientLoginInfo> {
        return activeClients[chid]?.values?.toList() ?: ArrayList()
    }

    fun getAllClients(): ArrayList<PushClientsManager.ClientLoginInfo> {
        return ArrayList<PushClientsManager.ClientLoginInfo>().apply {
            activeClients.values.forEach { addAll(it.values) }
        }
    }

    fun getClientLoginInfoByChidAndUserId(
        chid: String,
        userId: String,
    ): PushClientsManager.ClientLoginInfo? {
        return activeClients[chid]?.get(smtpLocalPart(userId))
    }

    fun getActiveClientMaps(): Iterable<HashMap<String?, PushClientsManager.ClientLoginInfo>> {
        return activeClients.values
    }

    fun removeActiveClients() {
        getAllClients().forEach { it.unwatch() }
        activeClients.clear()
    }

    fun removeAllClientChangeListeners() {
        clientChangeListeners.clear()
    }

    fun removeClientChangeListener(clientChangeListener: PushClientsManager.ClientChangeListener) {
        clientChangeListeners.remove(clientChangeListener)
    }

    private fun notifyListeners() {
        clientChangeListeners.forEach { it.onChange() }
    }

    private fun smtpLocalPart(userId: String?): String? {
        if (userId.isNullOrEmpty()) return null
        val atIndex = userId.indexOf('@')
        return if (atIndex > 0) userId.substring(0, atIndex) else userId
    }
}
